tests/PHPStan/Analyser/nsrt/bug-11716.php

```diff
@@ -0,0 +1,138 @@
+<?php declare(strict_types=1);
+
+namespace Bug11716;
+
+use function PHPStan\Testing\assertType;
+
+class TypeExpression
+{
+	/**
+	 * @return '&'|'|'
+	 */
+	public function parse(string $glue): string
+	{
+		$seenGlues = ['|' => false, '&' => false];
+
+		assertType("array{|: false, &: false}", $seenGlues);
+
+		if ($glue !== '') {
+			assertType('non-empty-string', $glue);
+
+			\assert(isset($seenGlues[$glue]));
+			$seenGlues[$glue] = true;
+
+			assertType("'&'|'|'", $glue);
+			assertType('array{|: bool, &: bool}', $seenGlues);
+		} else {
+			assertType("''", $glue);
+		}
+
+		assertType("''|'&'|'|'", $glue);
+		assertType("array{|: bool, &: bool}", $seenGlues);
+
+		return array_key_first($seenGlues);
+	}
+}
+
+/**
+ * @param array<int, string> $arr
+ * @param array<1|'2', string> $numericKeyArray
+ */
+function narrowKey($mixed, string $s, int $i, array $generalArr, array $arr, array $numericKeyArray): void {
+	if (isset($generalArr[$mixed])) {
+		assertType('mixed~array|object|resource', $mixed);
```

Comment:
Not directly related, but I wonder if substraction should be further and always generalized if possible, `mixed~type` can be always generalized, ie. into `string|int|float|bool|null` here.

Reply:
I don't think it would have a functional difference. or do you have something in mind which would improve?

Reply:
Pure readability of the type.